### PR TITLE
Prototype gz header migration

### PR DESCRIPTION
**Pull request by scpeters**
# 🎉 New feature

Prototype part of https://github.com/ignition-tooling/release-tools/issues/698

## Summary

The big diff is not so easy to read, so I recommend going commit-by-commit:

* https://github.com/ignitionrobotics/ign-utils/commit/b2e10989c84dc238af8a89d9768514d5dcc84b10: move the current `ignition` header folder to `gz`
* https://github.com/ignitionrobotics/ign-utils/commit/7c6e7e82a504cbab9c257cd0cd20ac9da20f61ba: make several changes needed to support the `gz` headers
*  https://github.com/ignitionrobotics/ign-utils/commit/6fb3171bd3600feed20cd0528fd220797dcb5d4a: add dummy `ignition/*` headers that simply redirect to their `gz/` counterpart

This takes a very minimal approach, just to change the include path without changing any namespaces or even the `IGNITION_*` variables in config.hh. The only macros changed are for header guards of the new `gz/*` headers.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by chapulina — COMMENTED** [review]
Thanks, @scpeters , this approach looks good to me. Some questions / comments:

* How do you think we should do the deprecation warnings? Using a `#pragma message()` in every `ignition` header?
* Did you use a script or just prototyped it manually?
* It would be good to move the files from `ignition` to `gz` so we keep their commit history.
* When this PR is ready for merging, we should update examples and tests

**Comment by scpeters** [comment]
> * How do you think we should do the deprecation warnings? Using a `#pragma message()` in every `ignition` header?

I'm not sure if `#pragma message()` works with gcc. I think we could define a macro that uses different directives depending on the compiler and include that in each `ignition/*` header just before the redirect include of the corresponding `gz/` header.

> * Did you use a script or just prototyped it manually?

I did this manually, but it shouldn't be too hard to script

> * It would be good to move the files from `ignition` to `gz` so we keep their commit history.

I did a `git mv` command in https://github.com/ignitionrobotics/ign-utils/pull/51/commits/b2e10989c84dc238af8a89d9768514d5dcc84b10 and you can confirm with a `git blame` that the history is still there:

* https://github.com/ignitionrobotics/ign-utils/blame/6fb3171bd3600feed20cd0528fd220797dcb5d4a/include/gz/utils/Environment.hh

It doesn't show up because github is a little confused, but I think our history is preserved with this approach.

> * When this PR is ready for merging, we should update examples and tests

we can decide how to split the work up into phases, but I believe this approach can be merged independently without breaking anything. We could optionally update the `#include <ignition/*>` calls in `src`, `test`, and `examples` as you say, but I was trying to see what's the minimal amount that needs to be changed for this first PR.

Some possible steps:

* Move header files to `gz/*` and add redirect headers to `ignition/*` (included in this PR)
* Replace all `#include <ignition/*>` statements in header files with `#include <gz/*>` (included in this PR)
* Replace all `#include <ignition/*>` statements in `src`, `test`, and `examples` source code with `#include <gz/*>` (not included in this PR)
* Replace all `#include <ignition/*>` statements in downstream code
* Change ign-cmake3's default value of `PROJECT_INCLUDE_DIR` from `ignition/${IGN_DESIGNATION}` to `gz/${IGN_DESIGNATION}` (or use `${GZ_DESIGNATION}` if that has been defined) in [IgnConfigureProject.cmake](https://github.com/ignitionrobotics/ign-cmake/blob/main/cmake/IgnConfigureProject.cmake#L91)
* Remove the `REPLACE_IGNITION_INCLUDE_PATH` parameter override once its no longer needed
* Add `using namespace ignition` to the `ignition/` headers
* Add portable `#warning` equivalent to `ignition/` headers

Because GitHub gets confused by the diff in this pull request, I like making a narrow PR like this to add the `gz/` headers in a backwards-compatible way. This allows flexibility in deciding when to change the c++ namespaces and add deprecation warnings

**Comment by scpeters** [comment]
we could arguably backport this to fortress since I think it's backwards compatible, but I'm not sure if it's worth it

**Comment by chapulina** [comment]
> we could arguably backport this to fortress

Something to keep in mind for later 👍🏽

File: include/ignition/utils.hh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <gz/utils.hh>
```
Comment (methylDragon) [inline include/ignition/utils.hh:18]: Is this meant to be `gz/utils/utils.hh`?

Comment (scpeters) [reply include/ignition/utils.hh:18]: in general, it should only have one `utils`, so `gz/utils.hh` is correct (look at the installed paths of any `ign-*` package to confirm)

the one exception is libsdformat, which has `sdf/sdf.hh` but that one is a bit different anyway

**Comment by chapulina** [comment]
> I'm not sure if #pragma message() works with gcc.

It prints a message for me, not sure what's missing? I think the only downside to it is that it's not an actual warning.

> I did a git mv command in https://github.com/ignitionrobotics/ign-utils/commit/b2e10989c84dc238af8a89d9768514d5dcc84b10 and you can confirm with a git blame that the history is still there:

Ah perfect, indeed GitHub gets confused in the PR diff

> we can decide how to split the work up into phases

I'm fine with that 👍🏽 We just have to make sure we don't introduce warnings in the intermediate steps.

**Comment by scpeters** [comment]
#53 was merged instead

**Pull request closed without merging**